src/sdk/pynni/nni/compression/torch/pruning/amc/amc_pruner.py

```diff
@@ -0,0 +1,325 @@
+
+        self.env_args = Namespace(
+            model_type=model_type,
+            preserve_ratio=preserve_ratio,
+            lbound=lbound,
+            rbound=rbound,
+            reward=reward,
+            n_calibration_batches=n_calibration_batches,
+            n_points_per_layer=n_points_per_layer,
+            channel_round=channel_round,
+            output=self.output
+        )
+
+        self.env = ChannelPruningEnv(
+            self, evaluator, val_loader, checkpoint, args=self.env_args)
+
+        if not self.export:
+            print('=> Saving logs to {}'.format(self.output))
+            self.tfwriter = SummaryWriter(logdir=self.output)
+            self.text_writer = open(os.path.join(self.output, 'log.txt'), 'w')
+            print('=> Output path: {}...'.format(self.output))
+
+            nb_states = self.env.layer_embedding.shape[1]
+            nb_actions = 1  # just 1 action here
+
+            rmsize = rmsize * len(self.env.prunable_idx)  # for each layer
+            print('** Actual replay buffer size: {}'.format(rmsize))
+
+            self.ddpg_args = Namespace(
+                hidden1=hidden1,
+                hidden2=hidden2,
+                lr_c=lr_c,
+                lr_a=lr_a,
+                warmup=warmup,
+                discount=discount,
+                bsize=bsize,
+                rmsize=rmsize,
+                window_length=window_length,
+                tau=tau,
+                init_delta=init_delta,
+                delta_decay=delta_decay,
+                max_episode_length=max_episode_length,
+                debug=debug,
+                train_episode=train_episode,
+                epsilon=epsilon
+            )
+            self.agent = DDPG(nb_states, nb_actions, self.ddpg_args)
+
+
+    def compress(self):
+        if not self.export:
+            self.train(self.ddpg_args.train_episode, self.agent, self.env, self.output)
+        self.export_pruned_model()
+
+    def train(self, num_episode, agent, env, output):
+        agent.is_training = True
+        step = episode = episode_steps = 0
+        episode_reward = 0.
+        observation = None
+        T = []  # trajectory
+        while episode < num_episode:  # counting based on episode
+            # reset if it is the start of episode
+            if observation is None:
+                observation = deepcopy(env.reset())
+                agent.reset(observation)
+
+            # agent pick action ...
+            if episode <= self.ddpg_args.warmup:
+                action = agent.random_action()
+                # action = sample_from_truncated_normal_distribution(lower=0., upper=1., mu=env.preserve_ratio, sigma=0.5)
+            else:
+                action = agent.select_action(observation, episode=episode)
+
+            # env response with next_observation, reward, terminate_info
+            observation2, reward, done, info = env.step(action)
+            observation2 = deepcopy(observation2)
```

Comment:
what is the meaning of this line?

Reply:
Good catch, this is from amc code, it seems not necessary to do the copy twice.

Reply:
Fixed.